src/coreclr/System.Private.CoreLib/src/System/TypedReference.CoreCLR.cs

```diff
@@ -16,6 +19,35 @@ public ref partial struct TypedReference
         private readonly ref byte _value;
         private readonly IntPtr _type;
 
+        private TypedReference(ref byte target, RuntimeType type)
+        {
+            _value = ref target;
+            _type = type.GetUnderlyingNativeHandle();
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static unsafe TypedReference MakeTypedReference(ref byte target, RuntimeType lastFieldType)
+            => new TypedReference(ref target, lastFieldType);
+
+        private static unsafe ref byte GetFieldDataReference(object target, RuntimeFieldInfo field)
+        {
+            ByteRef offset = default;
+            GetFieldDataReference(((RtFieldInfo)field).GetFieldDesc(), ObjectHandleOnStack.Create(ref target), offset.Create());
+            Debug.Assert(!Unsafe.IsNullRef(ref offset.Get()));
+            GC.KeepAlive(field);
+            return ref offset.Get();
+        }
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "TypedReference_GetFieldDataReference")]
+        private static unsafe partial void GetFieldDataReference(IntPtr fieldDesc, ObjectHandleOnStack target, ByteRefOnStack offset);
+
+        private static ref byte GetFieldDataReference(ref byte target, RuntimeFieldInfo field)
+        {
+            Debug.Assert(!Unsafe.IsNullRef(ref target));
+            int offset = RuntimeFieldHandle.GetInstanceFieldOffset((RtFieldInfo)field);
+            return ref Unsafe.AddByteOffset(ref target, offset);
+        }
```

Comment:
I will reconcile this implementation with Mono after I get closer to a green PR build.